sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java

```diff
@@ -1406,8 +1406,16 @@ public boolean apply(@Nonnull String input) {
                       klass, entry.getKey(), closestMatches));
           }
         }
-
+        // We need to pull the method directly from klass, to retain generic information.
         Method method = propertyNamesToGetters.get(entry.getKey());
+        String methodName = method.getName();
+        try {
+          method = klass.getMethod(methodName);
+        } catch (NoSuchMethodException e) {
+          // If the method is not found, fall back to just using the registered value.  This
```

Comment:
Two questions -- (1) How far upstream are you suggesting we do this?  It seems like we have the type information here specifically because of the klass parameter.  (2) How should we handle the case where the method isn't found, as outlined in the comment?